src/librustc_typeck/check/mod.rs

```diff
@@ -4426,7 +4431,12 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
                     instance,
                     promoted: None
                 };
-                let count = tcx.const_eval(param_env.and(global_id));
+
+                let count = if self.const_param_def_id(count).is_some() {
+                    Ok(self.to_const(count, self.tcx.type_of(count_def_id)))
+                } else {
+                    tcx.const_eval(param_env.and(global_id))
```

Comment:
It's probably clearer to move most of the variables above into this branch, as the first branch only uses `count_def_id`.

Reply:
Fixed, thanks!